Added AreaNthSelector

fedorkotov opened the pull request:
Added AreaNthSelector that is useful for nested features selection. Especially to select one of coplanar nested wires for subsequent extrusion, cutting or filleting.

This selector was developed as a result of discussion in #684

fedorkotov commented:
Examples for Solids and Shells are a bit contrived. Is there a way to get more than one Solid or Shell to select from while using Workplane? Or maybe using this selector for Solids and Shells makes no sense and support for them should be removed?

fedorkotov commented:
Black is complaining about formatting. IMHO it makes the code **less readable** by merging long sequences of fluent API calls that I format as one call per line into long lines with too many calls. But if that is the accepted formatting style of the project I will implement black's recommendations

marcus7070 commented:
> Black is complaining about formatting. IMHO it makes the code **less readable** by merging long sequences of fluent API calls that I format as one call per line into long lines with too many calls. But if that is the accepted formatting style of the project I will implement black's recommendations

Yes please @fedorkotov. I know black isn't perfect, but it's much easier to read one consistent style.

fedorkotov commented:
Two questions
1. `AreaNthSelector` is very similar to `LengthNthSelector` proposed in #646. Maybe I should add length selector in this pull request? 
2. Maybe I should add `A` (area) and `L` (length) should be added to string selectors grammar with allowed usage like this
   1. `>A`  - objects with the largest area
   2. `<L[-2]`  - objects with second smallest length

Or is it better to keep things separated and both of these additional features should have their separate pull requests?

marcus7070 commented:
> 1. `AreaNthSelector` is very similar to `LengthNthSelector` proposed in #646.

I was thinking the same thing so I actually did #646 earlier today, thinking that the process would help me review this PR. I just put up the code as #690.

> 2. Maybe I should add `A` (area) and `L` (length) to string selectors grammar with allowed usage like this
>    
>    1. `>A`  - objects with the largest area
>    2. `<L[-2]`  - objects with second smallest length
> 
> Or is it better to keep things separated and both of these additional features should have their separate pull requests?

In this case I think we should do another PR, because we can roll area, length and radius selectors into one review.

fedorkotov commented:
> I was thinking the same thing so I actually did #646 earlier today, thinking that the process would help me review this PR. I just put up the code as #690.

Ok. I will include LengthNthSelector into my pull request.
But first one more question. I see in your `LengthNthSelector` you use `isinstance(obj, (...))` to determine what kind of `Shape` was passed into `_NthSelector.key(..)` implementation. I used `Shape.ShapeType()` in my `AreaNthSelector` for the same purpose. I assumed that comparing strings would be faster than type information manipulation but did not measure that. Should I change to `isinstance`? Or are these variants roughly equivalent?

> In this case I think we should do another PR, because we can roll area, length and radius selectors into one review.

Ok.

marcus7070 reviewed (changes requested):
You can include the LengthNthSelector code from #690 in here if you want, or we can work on #690 after this gets merged (which would be less work for you), I don't mind.

marcus7070 commented:
> Unit tests for Solids and Shells are a bit contrived. Is there a way to get more than one Solid or Shell to select from while using Workplane?

Something like
```python
shell = (
    cq.Workplane()
    .box(1, 1, 1)
    .faces(">Z")
    .shell(0.1)
    .val()
)
workplane_shells = (
    cq.Workplane()
    .rarray(2, 2, 3, 4)
    .eachpoint(lambda loc: shell.located(loc))
)
```

You can also union different sized shells/solids, which would give you more variety to test an area selector on.

>  Or maybe using this selector for Solids and Shells makes no sense and support for them should be removed?

Solids and shells have clearly defined areas, so I support this selector working on them.

marcus7070 commented on cadquery/selectors.py:
```diff
@@ -454,6 +454,64 @@ def filter(self, objectlist: Sequence[Shape]) -> List[Shape]:
+
+    Or to create a lip on a case seam:
+        
+        cq.Workplane("XY")\
+            .rect(20, 20)\
+            .extrude(10)\
+            .edges("|Z or <Z")\
+            .fillet(2)\
+            .faces(">Z")\
+            .shell(2)\
+            .faces(">Z")\
+            .wires(AreaNthSelector(-1))\
+            .toPending()\
+            .workplane()\
+            .offset2D(-1)\
+            .extrude(1)\
+            .faces(">Z[-2]")\
+            .wires(AreaNthSelector(0))\
+            .toPending()\
+            .workplane()\
+            .cutBlind(2)    
```
A black compatible version here too please.

marcus7070 commented:
If you check https://raw.githubusercontent.com/CadQuery/cadquery/master/doc/classreference.rst you'll see where we need to add `AreaNthSelector` to make it appear in the list of selectors at https://cadquery--688.org.readthedocs.build/en/688/classreference.html#selector-classes

fedorkotov commented:
- Switched Shape.ShapeType() to isinstance to determine what kind of Shape was passed into _NthSelector.key(..) implementation in AreaNthSelector for consistency with other selectors based on _NthSelector
- Added LengthNthSelector developed by @marcus7070 in #690 with somewhat modified unit tests
- Added new selector references to documentation
- Black compatible code examples in AreaNthSelector docstring
- Explicitly marked _NthSelector class and _NthSelector.key(..) method as abstract using standard abc package


Can't figure out what caused Windows pipelines to fail. Don't have access to Windows machine right now.

The log says 
> DLL load failed: The specified module could not be found

But which module and which dll? And why?
Have you experienced something like that before?

marcus7070 reviewed (approved):
I think all our Windows CI are failing, it's not due to your code. Once we get that fixed and can see all the tests pass I think this looks good to merge.

Thanks for doing this @fedorkotov, there is a lot of work here especially the tests.

fedorkotov commented:
Forgot to check for non-planar or non-closed wires. Fixed. 
For that I catch all exceptions that  `Face.makeFromWires(..)` might throw. In general catching everything is a code smell. Should I catch something more specific?

adam-urbanczyk commented:
@fedorkotov I propose to leave the catch as-is for now. `makeFromWires` needs some refactoring anyhows.

marcus7070 reviewed (approved):
Good catch with that non-planar wire test.

marcus7070 commented:
Thanks @fedorkotov!
